tests/xbot/common/logic/LatchTest.java

```diff
@@ -0,0 +1,133 @@
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        assertEquals(latchTestObserver.getNumTimesUpdated(), 2);
+    }
+
+    @Test
+    public void testFallRiseRiseFallRiseFallObservingFallingEdge() {
+        Latch latch = new Latch(true, Latch.EdgeType.FallingEdge);
+        LatchTestObserver latchTestObserver = new LatchTestObserver();
+        latch.addObserver(latchTestObserver);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), null);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), null);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        assertEquals(latchTestObserver.getNumTimesUpdated(), 2);
+    }
+
+    @Test
+    public void testRiseFallRiseObservingBothEdges() {
+        Latch latch = new Latch(true, Latch.EdgeType.Both);
+        LatchTestObserver latchTestObserver = new LatchTestObserver();
+        latch.addObserver(latchTestObserver);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        latch.setValue(true);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.RisingEdge);
+        latch.setValue(false);
+        assertEquals(latchTestObserver.getLastUpdateEdgeType(), Latch.EdgeType.FallingEdge);
+        assertEquals(latchTestObserver.getNumTimesUpdated(), 3);
+    }
+}
+
+class LatchTestObserver implements Observer {
+
+    private Latch.EdgeType lastUpdateEdgeType = null;
+    private int numTimesUpdated = 0;
+
+    public Latch.EdgeType getLastUpdateEdgeType() {
+        return lastUpdateEdgeType;
+    }
+
+    public int getNumTimesUpdated() {
+        return numTimesUpdated;
+    }
+
+    @Override
+    public void update(Observable o, Object arg) {
+        numTimesUpdated++;
```

Comment:
Can we assert here that `arg` is an `EdgeType`? It'll throw a `ClassCastException` anyway, but it's easier to track down an assertion failure than a CCE.

Reply:
So do (arg instanceof EdgeType) here?

Reply:
Yeah, and I think you can include a message too:

```java
assertTrue(arg instanceof EdgeType, "Latch must provide an argument of type EdgeType");
```

Reply:
Added